src/mame/machine/k573mcr.h

```diff
@@ -9,19 +9,67 @@
 
 #pragma once
 
+#include "bus/psx/ctlrport.h"
+#include "machine/jvsdev.h"
 
-
-DECLARE_DEVICE_TYPE(KONAMI_573_MEMORY_CARD_READER, k573mcr_device)
-
-class k573mcr_device : public device_t
+class k573mcr_device : public jvs_device
 {
 public:
+	template <typename T>
+	k573mcr_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock, T &&jvs_host_tag)
+		: k573mcr_device(mconfig, tag, owner, clock)
+	{
+		host.set_tag(std::forward<T>(jvs_host_tag));
+	}
+
 	k573mcr_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
 
+	virtual ioport_constructor device_input_ports() const override;
+
 protected:
+	template <uint8_t First> void set_port_tags() { }
+
 	virtual void device_start() override;
+	virtual void device_reset() override;
+	virtual void device_add_mconfig(machine_config &config) override;
 
 	virtual const tiny_rom_entry *device_rom_region() const override;
+
+	// JVS device overrides
+	virtual const char *device_id() override;
+	virtual uint8_t command_format_version() override;
+	virtual uint8_t jvs_standard_version() override;
+	virtual uint8_t comm_method_version() override;
+
+	virtual int device_handle_message(const uint8_t *send_buffer, uint32_t send_size, uint8_t *&recv_buffer) override;
+
+private:
+	enum {
+		MEMCARD_UNINITIALIZED = 0x0000, // Default value, also used after writing?
+		MEMCARD_ERROR         = 0x0002,
+		MEMCARD_UNAVAILABLE   = 0x0008, // Card is not inserted
+		MEMCARD_READING       = 0x0200, // Read request is executing
+		MEMCARD_WRITING       = 0x0400, // Write request is executing
+		MEMCARD_AVAILABLE     = 0x8000  // Can be combined with MEMCARD_READING and MEMCARD_WRITING for busy state
+	};
+
+	enum {
+		RAM_SIZE = 0x400000,
+		MEMCARD_BLOCK_SIZE = 128
+	};
+
+	uint8_t controller_port_send_byte(uint32_t port, uint8_t data);
+	bool pad_read(uint32_t port_no, uint8_t *output);
+	bool memcard_read(uint32_t port_no, uint16_t block_addr, uint8_t *output);
+	bool memcard_write(uint32_t port_no, uint16_t block_addr, uint8_t *input);
+
+	std::unique_ptr<uint8_t[]> m_ram;
+	uint16_t m_status;
+	bool m_is_memcard_initialized;
+
+	required_device<psx_controller_port_device> m_ports[2];
```

Comment:
`required_device_array` should be used rather than an array of `required_device`.